Usage of util.apply_parallel in tutorial_parallelization.rst (#6837)
scikit-image/scikit-image: doc/source/user_guide/tutorial_parallelization.rst

    %timeit comprehension_loop()

> ``apply_parallel`` is a function in ``skimage.util`` module that provides mapping function in parallel across an array.

Reviewer: I think the grammar in "that provides mapping function" is off? How about

```suggestion
:func:`skimage.util.apply_parallel` can map a function in parallel across an array.
```

This seems a lot more concise to me. By using the :func: role we can make Sphinx add a link to the documentation of the function.

> In this guide, it is used to apply the ``task`` function to each element of a 5D array(image) in parallel.

Author: I will do that. Thanks.

```
:func:

.. note::  

.. code-block:: 

```
Is the above the right arrangement?

> In this guide, it is used to apply the ``task`` function to each element of a 5D array(image) in parallel.

Reviewer: Yes, that's the correct arrangement. :)

.. code-block:: python

    from skimage.util.apply_parallel import apply_parallel
    def apply_parallel_loop():
        [apply_parallel(task, image, dtype=image.dtype,
                        mode="nearest", channel_axis=-1) for image in pics]

>     %timeit apply_parallel_loop()

Reviewer: Running this block after the previous ones returns a `IndexError: tuple index out of range`. Does that happen for you as well? Could you look into it?

<details>
<summary>Full traceback</summary>

```
Traceback (most recent call last):
  File "~/.local/lib/venv/skimagedev/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 3460, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-4-9287c6c86bfb>", line 6, in <module>
    get_ipython().run_line_magic('timeit', 'apply_parallel_loop()')
  File "~/.local/lib/venv/skimagedev/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 2369, in run_line_magic
    result = fn(*args, **kwargs)
  File "~/.local/lib/venv/skimagedev/lib/python3.10/site-packages/IPython/core/magics/execution.py", line 1163, in timeit
    time_number = timer.timeit(number)
  File "~/.local/lib/venv/skimagedev/lib/python3.10/site-packages/IPython/core/magics/execution.py", line 157, in timeit
    timing = self.inner(it, self.timer)
  File "<magic-timeit>", line 1, in inner
  File "<ipython-input-4-9287c6c86bfb>", line 3, in apply_parallel_loop
    [apply_parallel(task, image, dtype=image.dtype,
  File "<ipython-input-4-9287c6c86bfb>", line 3, in <listcomp>
    [apply_parallel(task, image, dtype=image.dtype,
  File "~/scikit-image/skimage/util/apply_parallel.py", line 194, in apply_parallel
    res = res.compute()
  File "~/.local/lib/venv/skimagedev/lib/python3.10/site-packages/dask/base.py", line 314, in compute
    (result,) = compute(self, traverse=False, **kwargs)
  File "~/.local/lib/venv/skimagedev/lib/python3.10/site-packages/dask/base.py", line 600, in compute
    return repack([f(r, *a) for r, (f, a) in zip(results, postcomputes)])
  File "~/.local/lib/venv/skimagedev/lib/python3.10/site-packages/dask/base.py", line 600, in <listcomp>
    return repack([f(r, *a) for r, (f, a) in zip(results, postcomputes)])
  File "~/.local/lib/venv/skimagedev/lib/python3.10/site-packages/dask/array/core.py", line 1281, in finalize
    return concatenate3(results)
  File "~/.local/lib/venv/skimagedev/lib/python3.10/site-packages/dask/array/core.py", line 5274, in concatenate3
    chunks = chunks_from_arrays(arrays)
  File "~/.local/lib/venv/skimagedev/lib/python3.10/site-packages/dask/array/core.py", line 5061, in chunks_from_arrays
    result.append(tuple(shape(deepfirst(a))[dim] for a in arrays))
  File "~/.local/lib/venv/skimagedev/lib/python3.10/site-packages/dask/array/core.py", line 5061, in <genexpr>
    result.append(tuple(shape(deepfirst(a))[dim] for a in arrays))
IndexError: tuple index out of range

```

</details>

Author: Hi @lagru, I looked into it. The error doesn't happen for me when I run the code block. Without the `timeit` method, returns a list of arrays of length 87. Same as the previous blocks.

Reviewer: Strange. When I checkout your branch and run 
```python
from skimage import data, color, util
from skimage.restoration import denoise_tv_chambolle
from skimage.feature import hog
def task(image):
    """
    Apply some functions and return an image.
    """
    image = denoise_tv_chambolle(image[0][0], weight=0.1, channel_axis=-1)
    fd, hog_image = hog(color.rgb2gray(image), orientations=8,
                        pixels_per_cell=(16, 16), cells_per_block=(1, 1),
                        visualize=True)
    return hog_image
# Prepare images
hubble = data.hubble_deep_field()
width = 10
pics = util.view_as_windows(hubble, (width, hubble.shape[1], hubble.shape[2]), step=width)
from skimage.util.apply_parallel import apply_parallel
def apply_parallel_loop():
    [apply_parallel(task, image, dtype=image.dtype,
                    mode="nearest", channel_axis=-1) for image in pics]
apply_parallel_loop()
```
(copied from the tutorial) I get the error above. How are you running this code block? Which dask version do you have installed?

Author: dask version: 2023.3.2

When I checkout my branch and run the code in a `main` branch, it returned no errors. Likewise in IPython environment.

Reviewer 2: No, I'm getting the same error! :grimacing:

Reviewer 2: @lagru I started dabbling with [apply_parallel](https://scikit-image.org/docs/stable/api/skimage.util.html#skimage.util.apply_parallel) thanks to your suggestion! I was getting the same `IndexError: tuple index out of range` (running other code) until I realized that the `function` in `apply_parallel(function, array, ...)` was expected to return an array with the *same shape* as `array` (the input image).

This doesn't seem to be the issue here... I'm currently running the [code](https://scikit-image.org/docs/stable/user_guide/tutorial_parallelization.html) in a remote setup (see below), and the `task(image)` call is failing for any `image` in `pics` with error:

    ValueError: The input image is too small given the values of pixels_per_cell and cells_per_block. It should have at least: 16 rows and 16 cols.

**Environment:**
Python 3.9.5
Dask version: 2024.6.2
NumPy version: 1.26.4
scikit-image version: 0.24.0

    %timeit comprehension_loop()

:func:`skimage.util.apply_parallel` can map a function in parallel across an array.
In this guide, it is used to apply the ``task`` function to each element of a 5D array(image) in parallel.

.. note::

>     ``apply_parallel`` needs the optional dependency dask to be installed.

Reviewer: ```suggestion
    :func:`skimage.util.apply_parallel` needs the optional dependency dask_ to be installed.
    
.. _dask: https://www.dask.org
```
